rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,82 @@
+| **Author(s)** | Pankaj Kanwar (Google), Reed Wanderman-Milne (Google)|
+| **Sponsor**   | Sanjoy Das (Google)                                  |
+| **Updated**   | 2021-01-31                                           |
+
+
+## Objective
+Allow users to enable deterministic behavior in TensorFlow. This means if the user runs a TensorFlow program multiple times, the model outputs and weights will be the same each time. Determinism will be supported on CPUs and GPUs.
+ 
+To get deterministic behavior, users must do the following:
+
+* Enable determinism using the API proposed in this doc.
+* Use same hardware in every run.
+* Use the same software environment every run (OS, checkpoints, version of TF, environmental variables, etc).
+* Not use constructs outside TensorFlow that are nondeterministic, such as Python’s `random` module or using multiple threads/processes in ways that influence TensorFlow’s behavior.
+* Do not use nondeterministic custom ops.
+
+## Motivation
+There are several mission critical applications in life sciences, finance and automation that require deterministic behavior. Determinism is required so that the behavior of these applications can be accurately predicted & demonstrated in a variety of scenarios.
+
+Lack of determinism prevents companies from launching products using models developed in TF. For a subset of these industries having deterministic behavior is a regulatory requirement. 
+
+In addition, determinism increases model velocity development by reducing noise, while also simplifying the debugging workflow.
+
+## Design Proposal
+We will create a new flag with the default value of ‘False’ which enables determinism.  We will define 2 functions:
+
+* `tf.config.enable_deterministic_execution(enabled)`
+* `tf.config.deterministic_execution_enabled()`
+
+The first function takes in a boolean value, and allows the model developer to enable/disable determinism. The second function returns a bool indicating whether determinism is enabled.
+In some cases, we have deterministic and nondeterministic versions of the kernel. In such cases, we will use this flag to run the appropriate kernels.
+For ops which do not yet have a deterministic implementation, TensorFlow will raise a `tf.errors.UnimplementedError` if the flag is enabled.
+
+Enabling deterministic execution does not automatically cause a user’s program to become deterministic. If users use nondeterministic constructs outside TensorFlow, such as threads/process, in ways that influence TensorFlow’s behavior, their program will not be deterministic. In order for a user to ensure their program is deterministic, users must both enable deterministic execution within TensorFlow and remove any sources of nondeterminism outside TensorFlow.
+
+### Existing Flags
+Multiple environmental variables exist today that control determinism. As part of this change, we will deprecate then remove the following:
+
+* TF_DETERMINISTIC_OPS
+* TF_CUDNN_DETERMINISTIC
```

Comment:
This comment addresses everything from "Multiple environmental variables ..." to "TF_CUDNN_DETERMINISTIC" (lines 41 to 44).

I think that this RFC should be more explicit about the current state of nondeterminism functionality in TensorFlow and how this API change reflects the next step in the process of getting to 100% determinism for all ops and all models.

I propose the following:

There are currently two environment variables in TensorFlow to enable deterministic op functionality.

The first environment variable is `TF_CUDNN_DETERMINISTIC`. When set to 'true' or '1', this,

  * makes the selection of cuDNN convolution algorithms deterministic,
  * selects deterministic gradient algorithms for `tf.nn.conv*d` and `tf.keras.layers.Conv*D`,
  * selects deterministic gradient algorithms for `tf.nn.max_pool*d` and `tf.keras.layers.MaxPool*D`, and
  * selects a deterministic gradient algorithm for `tf.nn.ctc_loss`.

The second environment variable is `TF_DETERMINISTIC_OPS`. This supercedes and replaces `TF_CUNN_DETERMINISTIC` by first implementing the same functionality, but then it also (when set to 'true' or '1'),

  * selects deterministic gradient kernels for `tf.nn.bias_add` and the many Keras layers that apply a bias,
  * selects a deterministic algorithm for XLA reductions on GPU, and
  * selects a deterministic gradient algorithm for `tf.image.resize` with `method=ResizeMethod.BILINEAR` and `tf.keras.layers.UpSampling2D` with `interpolation='bilinear'`
  
Calling `tf.config.enable_deterministic_execution(True)` will be equivalent to setting `TF_DETERMINISTIC_OPS` to 'true' or '1' plus the additional functionality described in this RFC.

The two environment variables will be first deprecated and then removed.